qr-code/src/main/java/com/adyen/checkout/qrcode/DefaultQRCodeDelegate.kt

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2022 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by oscars on 15/8/2022.
+ */
+
+package com.adyen.checkout.qrcode
+
+import androidx.annotation.VisibleForTesting
+import com.adyen.checkout.components.model.payments.response.Action
+import com.adyen.checkout.components.model.payments.response.QrCodeAction
+import com.adyen.checkout.components.status.DefaultStatusRepository
+import com.adyen.checkout.components.status.StatusRepository
+import com.adyen.checkout.components.status.api.StatusResponseUtils
+import com.adyen.checkout.components.status.model.StatusResponse
+import com.adyen.checkout.core.exception.CheckoutException
+import com.adyen.checkout.core.exception.ComponentException
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import org.json.JSONException
+import org.json.JSONObject
+import java.util.concurrent.TimeUnit
+
+@Suppress("TooManyFunctions")
+internal class DefaultQRCodeDelegate(
+    private val statusRepository: StatusRepository,
+    private val statusCountDownTimer: QRCodeCountDownTimer,
+) : QRCodeDelegate {
+
+    private val _outputDataFlow = MutableStateFlow<QRCodeOutputData?>(null)
+    override val outputDataFlow: Flow<QRCodeOutputData?> = _outputDataFlow
+
+    override val outputData: QRCodeOutputData? get() = _outputDataFlow.value
+
+    private val _exceptionFlow = MutableSharedFlow<CheckoutException>(0, 1, BufferOverflow.DROP_OLDEST)
+    override val exceptionFlow: Flow<CheckoutException> = _exceptionFlow
+
+    private val _detailsFlow = MutableStateFlow<JSONObject?>(null)
+    override val detailsFlow: Flow<JSONObject?> = _detailsFlow
+
+    private val _timerFlow = MutableStateFlow(TimerData(0, 0))
+    override val timerFlow: Flow<TimerData> = _timerFlow
+
+    private var _coroutineScope: CoroutineScope? = null
+    private val coroutineScope: CoroutineScope get() = requireNotNull(_coroutineScope)
+
+    private var statusPollingJob: Job? = null
+
+    private var qrCodeData: String? = null
+
+    init {
+        statusCountDownTimer.attach(
+            DefaultStatusRepository.MAX_POLLING_DURATION_MILLIS,
+            STATUS_POLLING_INTERVAL_MILLIS
+        ) { millisUntilFinished -> onTimerTick(millisUntilFinished) }
+    }
+
+    @VisibleForTesting
+    internal fun onTimerTick(millisUntilFinished: Long) {
+        val progressPercentage =
+            (HUNDRED * millisUntilFinished / DefaultStatusRepository.MAX_POLLING_DURATION_MILLIS).toInt()
```

Comment:
Should these constants be in the `StatusRepository` instead of the `DefaultStatusRepository`

Reply:
Agreed, but `DefaultQRCodeDelegate` doesn't know that we're injecting `DefaultStatusRepository` inside it and not another implementation of `StatusRepository`.

What if we add a method `getMaxPollingDurationMillis` to `StatusRepository` then? (this is all very minor btw, just discussing it)

Reply:
That's a good idea actually!